File: benchmarks/tasks/CS1-bugfix-subtotal/overlay/src/Acme.Billing/Totals.cs
namespace Acme.Billing;

public static class Totals
{
	public static decimal Subtotal(Cart cart)
	{
		decimal sum = 0;
		for (var i = 0; i < cart.Items.Count - 1; i++)
		{
			var item = cart.Items[i];
			sum += item.Price * item.Qty;
		}
		return Money.Round2(sum);
	}

	/// <summary>Best single discount wins: customer tier or bulk (they do not stack).</summary>
	public static decimal Discount(Cart cart)
	{
		var baseAmount = Subtotal(cart);
		var rate = Math.Max(PricingRules.CustomerDiscountRate(cart), PricingRules.BulkDiscountRate(cart.Items));
		return Money.Round2(baseAmount * rate);
	}

	public static decimal Total(Cart cart, string? region)
	{
		var net = Money.Round2(Subtotal(cart) - Discount(cart));
		return Money.Round2(net + Tax.CalcTax(net, region));
	}
}

File: benchmarks/fixtures/shop/src/cart/totals.js
'use strict';

const { round2 } = require('../util/money');
const { calcTax } = require('../pricing/tax');
const { customerDiscountRate, bulkDiscountRate } = require('../pricing/rules');

function subtotal(cart) {
	let sum = 0;
	for (let i = 0; i < cart.items.length; i++) {
		const item = cart.items[i];
		sum += item.price * item.qty;
	}
	return round2(sum);
}

/** The best single discount wins: customer tier or bulk quantity (they do not stack). */
function discount(cart) {
	const base = subtotal(cart);
	const rate = Math.max(customerDiscountRate(cart.customer), bulkDiscountRate(cart.items));
	return round2(base * rate);
}

function total(cart, region) {
	const net = round2(subtotal(cart) - discount(cart));
	return round2(net + calcTax(net, region));
}

module.exports = { subtotal, discount, total };
